src/main/kotlin/store/storerepository/StoreRepositoryPostgreSql.kt

```diff
@@ -0,0 +1,138 @@
+
+// TODO should be dealing with real entities only
+class StoreRepositoryPostgreSql(private val database: Database) : StoreRepository {
+
+    init {
+        transaction(database) {
+            SchemaUtils.createMissingTablesAndColumns(
+                StoreSchema,
+                StoreExtraFieldSchema,
+            )
+        }
+    }
+
+    private object StoreSchema : Table("stores") {
+        val id = varchar("id", 36).primaryKey()
+        val externalId = varchar("external_id", 20).uniqueIndex()
+        val name = varchar("name", 50).nullable()
+        val customName = bool("custom_name").nullable()
+        val description = varchar("description", 2000).nullable()
+        val type = varchar("type", 40).nullable()
+        val openingDate = date("opening_date").nullable()
+        val code = varchar("code", 70).nullable()
+        val seasons = blob("seasons_json").nullable()
+    }
+
+    private object StoreExtraFieldSchema : Table("store_extra_fields") {
+        val storeId = varchar("id", 20).references(StoreSchema.externalId, onDelete = CASCADE).primaryKey()
+        val name = varchar("name", 80).primaryKey()
+        val value = varchar("value", 150).nullable()
+    }
+
+    override fun list(page: Int?, nameSearch: String?) = transaction(database) {
+        StoreSchema.selectAll()
+            .orderBy(StoreSchema.id, SortOrder.DESC)
+            .apply {
+                page?.let {
+                    limit(10, page * 10)
+                }
+                nameSearch?.takeUnless(String::isBlank)?.let {
+                    adjustWhere {
+                        Op.build { StoreSchema.name.lowerCase() like "%$it%".toLowerCase() }
+                    }
+                }
+            }.map {
+                Store(
+                    id = it[StoreSchema.id],
+                    externalId = it[StoreSchema.externalId],
+                    name = it[StoreSchema.name],
+                    code = it[StoreSchema.code],
+                    description = it[StoreSchema.description],
+                    type = it[StoreSchema.type],
+                    openingDate = it[StoreSchema.openingDate]?.toDate(),
+                    extraFields = listExtraFields(it[StoreSchema.externalId]), // TODO use a JOIN instead
+                    seasons = it[StoreSchema.seasons]?.parseSeasons() ?: emptySet()
+                )
+            }
+    }
+
+    private fun Blob.parseSeasons() =
+        objectMapper.readValue(binaryStream, object : TypeReference<List<String>>() {}).toSet()
+
+    private fun listExtraFields(storeId: String) =
+        StoreExtraFieldSchema.select { StoreExtraFieldSchema.storeId eq storeId }
+            .map { it[StoreExtraFieldSchema.name] to it[StoreExtraFieldSchema.value] }
+            .toMap()
+
+    override fun saveInfo(storeInfo: StoreInfo) {
+        transaction(database) {
+            val store = StoreSchema.select { StoreSchema.externalId eq storeInfo.externalId }.singleOrNull()
```

Comment:
If performance was really an issue, how would make this upsertion more efficient?

Reply:
I could have used a [native query to PostgreSQL upsert](https://www.postgresqltutorial.com/postgresql-upsert/) so that, while inserting an update is done on a conflict.